File: 3016-minimum-number-of-pushes-to-type-word-ii/solution.cpp
#include <string>
#include <vector>
#include <algorithm>

class Solution {
public:
    int minimumPushes(std::string word) {
        // Frequency array for 26 lowercase English letters
        std::vector<int> freq(26, 0);
        for (char c : word) {
            freq[c - 'a']++;
        }
        
        // Sort frequencies in descending order
        std::sort(freq.rbegin(), freq.rend());
        
        int totalPushes = 0;
        for (int i = 0; i < 26; ++i) {
            if (freq[i] == 0) {
                break; // No more characters present in the string
            }
            // First 8 characters take 1 push, next 8 take 2 pushes, etc.
            int multiplier = (i / 8) + 1;
            totalPushes += freq[i] * multiplier;
        }
        
        return totalPushes;
    }
};

File: 3016-minimum-number-of-pushes-to-type-word-ii/solution.py
from collections import Counter

class Solution:
    def minimumPushes(self, word: str) -> int:
        # Count frequency of each character
        counts = Counter(word)
        
        # Sort frequencies in descending order
        sorted_freqs = sorted(counts.values(), reverse=True)
        
        total_pushes = 0
        # Calculate push cost based on rank (every 8 characters increment push cost)
        for i, freq in enumerate(sorted_freqs):
            multiplier = (i // 8) + 1
            total_pushes += freq * multiplier
            
        return total_pushes